Firestore/core/src/firebase/firestore/api/query_snapshot.mm

```diff
@@ -63,6 +67,91 @@ DocumentSnapshot snap(firestore_, document.key, document, from_cache,
   }
 }
 
+static DocumentChange::Type DocumentChangeTypeForChange(
+    const DocumentViewChange& change) {
+  switch (change.type()) {
+    case DocumentViewChange::Type::kAdded:
+      return DocumentChange::Type::Added;
+    case DocumentViewChange::Type::kModified:
+    case DocumentViewChange::Type::kMetadata:
+      return DocumentChange::Type::Modified;
+    case DocumentViewChange::Type::kRemoved:
+      return DocumentChange::Type::Removed;
+  }
+
+  HARD_FAIL("Unknown DocumentViewChange::Type: %s", change.type());
+}
+
+void QuerySnapshot::ForEachChange(
+    bool includeMetadataChanges,
+    const std::function<void(DocumentChange)>& callback) const {
+  if (includeMetadataChanges && snapshot_.excludes_metadata_changes()) {
+    ThrowInvalidArgument("To include metadata changes with your document "
+                         "changes, you must call "
+                         "addSnapshotListener(includeMetadataChanges:true).");
+  }
+
+  if (snapshot_.old_documents().empty()) {
+    // Special case the first snapshot because index calculation is easy and
+    // fast. Also all changes on the first snapshot are adds so there are also
+    // no metadata-only changes to filter out.
+    FSTDocument* lastDocument = nil;
+    size_t index = 0;
+    for (const DocumentViewChange& change : snapshot_.document_changes()) {
+      FSTDocument* doc = change.document();
+      SnapshotMetadata metadata = SnapshotMetadata(
+          /*hasPendingWrites=*/snapshot_.mutated_keys().contains(doc.key),
+          /*fromCache=*/snapshot_.from_cache());
+      DocumentSnapshot document =
+          DocumentSnapshot(firestore_, doc.key, doc, std::move(metadata));
+
+      HARD_ASSERT(change.type() == DocumentViewChange::Type::kAdded,
+                  "Invalid event type for first snapshot");
+      HARD_ASSERT(!lastDocument || snapshot_.query().comparator(
+                                       lastDocument, change.document()) ==
+                                       NSOrderedAscending,
+                  "Got added events in wrong order");
+
+      callback(DocumentChange(DocumentChange::Type::Added, document,
```

Comment:
As far as I can see, you can `std::move` `document` into the `DocumentChange` constructor (`document` is a local variable that doesn't seem to be used after this line).